### ISSUE TITLE
Version control on package dependency

### ISSUE THREAD

**Issue by zhengle-advantaseeds** [ficonsulting/RInno#48]
I am facing some issues caused by the dependency version. For example:
1. I built the app using RInno with R 3.4.2 last year and distributed to my colleagues. 
2. The dependency libraries worked fine because of the up-to-date R, like ggplot2
3. However, after 3.4.3 released, when I deployed the app (build on R 3.4.2), it may not be able to start-up correctly.
4. I checked the log: it seems that some dependencies (like labeling) of ggplot2 cannot be correctly installed automatically. But, I can install it using R interface, and the problem solved.

Like in Deploy R project, the portable app will carry its own dependencies. Would it be a good idea to integrate it in the RInno? For example, using Packrat.

**Comment by chasemc** [comment]
I have faced similar issues and am in the process of tracing the the source (I'm thinking it is the new RInno package management but cannot say for sure).

I am also working on implementing PackRat, with my only concern being the time to compile the packages which will make installation more cumbersome. 

I'm busy with other projects at the moment, but your users will need Rtools to use Packrat.  I have started working on this by modifying the code.iss as below:

https://gist.github.com/chasemc/cbb7f1cca4d261890096a9da2f05264d

**Comment by Dripdrop12** [comment]
The old methodology was agnostic to the R package version. It just checked to see if a given package was installed or not (T/F). If the package had not been installed, then `install.packages` was used to install the most recent CRAN version of that package.

The new methodology checks for a version greater than or equal to the version used in development (this can also be controlled with manually defined inequalities). If that condition is not met, then it installs the version provided in the conditional. There is a more precise check and a more precise action following that check.

In some situations, RTools & `packrat` might be an option. So I really like where @chasemc is headed with his gist. However, `packrat` is not ideal for a few reasons:
1. It is designed to help make R projects more reproducible, not manage the software dependencies of a shiny app.
2. It is highly automated and difficult to customize to new uses
3. I spent about 3 weeks working on it before changing tracks towards the current solution 

I'm not saying `packrat` is a dead end, but we should be able mimic most of its functionality already by specifying your dependencies more precisely like this:

```
create_app(
  app_name = "myapp", 
  app_dir = "app",
  pkgs = c(shiny = "==1.0.5", dplyr = "==0.7.4")
  include_R = TRUE,
  R_version = "==3.4.3"
)
```

**Comment by chasemc** [comment]
Agreed,  which is why I'll share what I wind up doing, but probably won't submit a PR.  I am using it for a bioinformatics tool, so I need the results to be reproducible. My current issue would probably be  making sure users have the correct version of dependencies of dependencies (hence packrat).

**Comment by Dripdrop12** [comment]
@chasemc do you think we can add `dependencies = TRUE` to get your desired result? We are already using `devtools::install_version` which defaults to upgrading out of date dependencies of dependencies. 

See `?devtools::install` for details.

**Comment by georgemirandajr** [comment]
I use R at work and there appears to be limitations when it comes to downloading packages and other things involving an internet connection. For instance, when I used RInno the first time, my app could not download the required packages during initialization (i.e., opening the app right after the install). So my workaround is to bundle my app with a library folder that I already put the needed packages in. This also is a handy way for me to ensure that the package dependencies remain intact. 

This seemed to work really well except for one thing: the package `data.table` seems to lose the `.` in its folder name when it installs (pic below). **Do you think this is an issue with the compression/decompression?** When I add the `.` back in, the app loads fine and works beautifully. You can imagine me shaking my fist in the air at the creator of data.table for making a wonderful package with an unfortunate name. 

![image](https://user-images.githubusercontent.com/6701264/36159379-6989519c-1093-11e8-8455-4ff99849f2c1.png)

**Comment by Dripdrop12** [comment]
For the `.` issue, @chasemc fixed it: fe7f89bed94995cd963e1205a865e5f23015760b.

We also added support for `locals` and `include_Rtools` which will allow you to build locally supplied packages on users' computers. I think this solution is more stable because copying installed packages could create strange architecture bugs among other things. 

I recommend that you pull down the source code of your shiny app's dependencies. Then use `devtools`:

```
devtools::build(pkg = "path/to/dependency/source/code", path = "app_dir/local")
```
In `create_app`, you can use the `locals` argument to build the packages from source and `include_Rtools` to install Rtools with your app:

```
create_app(
  app_name = "myapp",
  locals = c("dplyr", "shiny", "jsonlite", "rmarkdown"),
  include_Rtools = TRUE
)
```
This is a more reliable way to provide an all-in-one, stable application. I think that addresses this issue, but we welcome further discussion about how to improve from here!

**Comment by r2evans** [comment]
Late to the conversation, @georgemirandajr , but would using `checkpoint` (and MRAN) work for you? That would force packages current as of a particular day, which could reduce deployment liabilities for you.

**Comment by georgemirandajr** [comment]
Not too familiar with checkpoint but it appears that it will address the
issue I was having.

**Comment by r2evans** [comment]
In my experience, the difference between `packrat` (and now `renv`, replacing it) and `checkpoint` is about two things: the supporting infrastructure (some choose to avoid MRAN for perhaps anti-MS sentiments) and problems with development workflow. For the latter, when I am developing a project for a client, I'll already have the basics installed (like `ggplot2`), but when I need to add a new package two months into development, I'll `install.packages` it from that day but not necessarily update all of the other packages used for that project. Because of this, a single point-in-time is often not perfectly representative of the state when I wrote the package, `renv` is a better bet.


I honestly think `renv` (`packrat`) is ultimately a better path to go in general.

While working on a project for a client, I'll already have some basic packages installed (e.g., `dplyr`), but when I need to add a new (to me) package, I'll just `install.packages` that one package; if I don't update all other installed packages, then using "today" (the day of installing the newer packages) might not work, since it's possible (even likely) that `dplyr` had another update in the time between starting the project and now.

By using `checkpoint`'s "date" as the marker for which package versions to use, this means that either I have to find a date where all of my current versions are all the *most recent* versions (likely to fail), or whenever I install or update one package, I have to update all packages. (This can easily over-tax the testing processes.)

Using `renv` (or `packrat`), I can specify an arbitrary combination of versions, where by "arbitrary" I mean it doesn't matter when a package-version was released, I'm still using it. Similar to `python`'s `requirements.txt` file for `pip` package management.

\*shrug\*
